### PR TITLE
Sort and Merge Classes in WebCmdlets\Common

### PR DESCRIPTION
**Pull request by markekraus**
## PR Summary

Reference #5610

* Merges partial classes in the WebCmdlets\Common folder
* Sorts the `using` statements
* Sorts class elements according to StyleCop guidance https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1201.md
* Sorts within the same elements by access type according to StyleCop guidance https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1202.md
* Sorts elements of the same type and access alphabetically.
* Order of cmdlet public properties and parameters was untouched to preserve their order in PowerShell
* Proper inheritance was added to `InvokeRestMethodCommand`.
* Some new regions were created where it makes sense
* Formatting and comments of code blocks was purposefully maintained. These will be handled in a separate cleanup action.
* No logic was modified

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- NA User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - NA Issue filed - Issue link:
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- NA [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [x] [Add `[feature]` if the change is significant or affectes feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/ContentHelper.Common.cs [hunk 1]
```diff
@@ -26,22 +26,22 @@ internal static partial class ContentHelper
...
+        internal static Encoding GetEncoding(HttpResponseMessage response)
         {
-            contentType = GetContentTypeSignature(contentType);
-            return CheckIsJson(contentType);
+            // ContentType may not exist in response header.
+            string charSet = response.Content.Headers.ContentType?.CharSet;
+            return GetEncodingOrDefault(charSet);
         }
 
         internal static Encoding GetEncodingOrDefault(string characterSet)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/ContentHelper.Common.cs]: @markekraus We could consider to change the default to Utf8 in later PR.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 2]
```diff
@@ -3,28 +3,49 @@
...
-using System.Net.Http;
-using System.Net.Http.Headers;
+using System.Text;
+using System.Text.RegularExpressions;
 using System.Threading;
 using System.Xml;
-using System.Collections.Generic;
-using System.Text.RegularExpressions;
-using System.Linq;
+using Microsoft.Win32;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: Do we really need this?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs [hunk 3]
```diff
@@ -3,11 +3,11 @@
 --********************************************************************/
 
 using System;
+using System.Collections.Generic;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs]: The file is not sorted but I agree to merge "as is" - if we'll have follow formatting PR.

**Review by iSazonov — APPROVED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs (hunk 3 above)
Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs]: @iSazonov I'm confused. This file is sorted. What is out of order?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 2 above)
Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: It looks like we don't.  Should I remove it in this PR or in another?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: Better in next - the PR is very large :-)

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs (hunk 3 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs]: Line 67.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 2 above)
Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: I have added it to the task list in #5610

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs (hunk 3 above)
Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs]: ah. that one can probably be fixed with automatic property `{ get; private set }`, which is on the checklist already.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs]: @markekraus Pow("Please", 10) make the follow PRs as small as possible - split work by file/type formattings. :-)

Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs]: @iSazonov if you have ideas on how the work should be split, please update #5610. I still don't have a good handle on what is too big or too small and, honestly, sometimes the feedback on that is inconsistent.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs]: Maybe interesting https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/

**Comment by TravisEz13** [comment]
1. If you have enabled styleCop, can you enable https://www.codefactor.io on your branch so I can see that your changes improve the style?
1. Such a large amount of code was changed, I have to take your word that no logic was changed.  Can we break this up into smaller PRs to get this done?   The following are ideas on how to break it up:
    - Perhaps merge the partial classes without sorting ( should be easy to review )
    - Sort using statements (should be easy to review)
    - Sort various element attempting to keep under a line limit (I would complete at least one file.)  The link [@iSazonov](https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/) suggested says 400.
    - Sometimes between coworkers, we use commits for breaking up reviews

**Comment by markekraus** [comment]
@TravisEz13 I actually haven't enabled stylecop. I tried but couldn't get it to work. I believe it would apply to the entire utility assembly which would make it very hard to track down just the web cmdlet code issues. This was all somewhat beyond me. So I resorted to sorting the code manually and using the stylecop rules as guidance. We can't use them all as, for example, our contributor guidelines say to group the private backing fields with public properties. that will make it hard to see what private field  and private-before-public sorting issues are actually issues.

I can break it up, yes. I tried breaking up a large change into the multiple commits in a single PR before, but @iSazonov didn't like that. I don't care how the work is broken up, but I could use some guidance, clarity, and consistency. I can be told one thing in a PR/issue, follow that to the letter in a new PR, and then have reviewers telling me to redo it a different way. This pattern has repeated several times and results in a lot of wasted time for all involved.

So, should I break this out into separate PRs? Or should I do multiple commits in a single PR or should I group a few commits in a few PRs? Should I move the files first? start with the partial class merges?

**Comment by TravisEz13** [comment]
@markekraus Can you point me to the previous PR with the objection from @iSazonov ?

**Comment by markekraus** [comment]
https://github.com/PowerShell/PowerShell/pull/5518#issuecomment-346425873

**Comment by TravisEz13** [comment]
I think we should try to take an incremental approach.  One of the reasons to enable codefactor.io on your branch (maybe we should enable it after 6.0.0 is released) is to show that the PR was an overall improvement (did not introduce more issues than in removed,) not to say that you have to fix every issue.   

I reviewed this link (https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/) further and I agree with the line limit and the reason behind it.  I spend an hour reviewing this PR and completed a few hundred lines.

Is there a reason that we cannot do a simple merge of the partial classes without any sorting (perhaps you could sort the using statements?)  This seems like a good starting point.

**Comment by markekraus** [comment]
Right now the pieces of the partial classes are in the same file. What constitutes a simple merge in this case? do I just move that code so it's in the same class brackets at the end of the class? 

One reason why I was merging and sorting at the same time is that the classes are so out of alignment it's hard to even chose a logical place to move the partial code into. I can move the code directly in the end of the larger class, but then that code is all unsorted and that's sort of how it is now, only its in its own partial class block. If i try to move it with similar element types, I might have 2 to 3 places to chose from. If i try to move it to where it _should_ be it will be out of place because the code around it is not where it should be.

**Comment by TravisEz13** [comment]
Another approach would be to focus on a limited number of classes to keep under the line limit.  I see one file that is over the line limit by itself.  I'd say try this approach with the smaller files and we can proceed from there.  @daxian-dbw @adityapatwardhan We should update the guidelines, but there should not be a requirement that a certain change is completed in a single PR unless splitting the change would have a negative effect on the project.

**Comment by markekraus** [comment]
Ok. I'm going to close this PR so I can keep the branch for reference. I turned on codefactor.io, but I'm not real happy with the full read/write access it requires. I will work on smaller chunks per PR.

**Comment by iSazonov** [comment]
@TravisEz13 Good news (although preview only) https://developer.github.com/v3/enterprise-admin/pre_receive_hooks/
Another way is GitHub Webhooks https://developer.github.com/v3/orgs/hooks/

**Pull request closed without merging**